### ISSUE TITLE
ConvertFrom-Json returns array with extra null element, if comments are outside the root element.

### ISSUE THREAD

**Issue by mubed** [PowerShell/PowerShell#21203]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://learn.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

Having a json file containing this:

```
// Comment on top level
{
    "key": "value"
}
```

should become a PSObject or Hashtable, when imported using Get-Content and then piped through ConvertFrom-Json, but the result is an array with a null as first element and the PSObject/Hashtable as the second element!

### Expected behavior

```console
Comment lines have to be ignored, not be included as null elements.
```

### Actual behavior

```console
Comments on top level are interpreted as null elements like a "{}".
```

### Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.4.1
PSEdition                      Core
GitCommitId                    7.4.1
OS                             Microsoft Windows 10.0.19045
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by mubed** [comment]
Might be related to https://github.com/PowerShell/PowerShell/issues/14553

**Comment by mubed** [comment]
Currect workaround:

`Get-Content File.json | Out-String | ConvertFrom-Json`

**Comment by rhubarb-geek-nz** [comment]
>> Having a json file containing this:

What you have there is a thing called _not a json file_. 

Comments are not part of the [JSON file specification](https://www.json.org/json-en.html).

Relying on comments in what you think are json files is just storing up trouble depending on the strictness of the parser you use. The simplest approach is to (a) stick to the JSON specification (b) use text elements within the structure as your comments.

**Comment by mubed** [comment]
> > > Having a json file containing this:
> 
> What you have there is a thing called _not a json file_.
> 
> Comments are not part of the [JSON file specification](https://www.json.org/json-en.html).
> 
> Relying on comments in what you think are json files is just storing up trouble depending on the strictness of the parser you use. The simplest approach is to (a) stick to the JSON specification (b) use text elements within the structure as your comments.

I would appreciate it, if you would first check the specifications of ComvertFrom-Json before writing google-able general information about common knowledge like what exactly JSON is:

https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/convertfrom-json?view=powershell-7.4

**Comment by rhubarb-geek-nz** [comment]
PowerShell is moving from Newtonsoft.Json to System.Text.Json and the grammer rules and capabilities are not the same for the extensions to JSON.

**Comment by rhubarb-geek-nz** [comment]
I see the null if you try a round trip. So it does not appear in the formal table output.

```
PS> Get-Content -LiteralPath file.json | ConvertFrom-Json | ConvertTo-Json
[
  null,
  {
    "key": "value"
  }
]
```

**Comment by 237dmitry** [comment]
> and the Out-String solves the problem

Yes, `null` in output, converting to string solves the problem:
```
Get-Content file.json -Raw | ConvertFrom-Json | ConvertTo-Json
```

**Comment by mubed** [comment]
I already mentioned above that there is a workaround using a single string (out-string or gc -raw) but this is not the **solution**. According to the documentation of ConvertFrom-Json, the current implementation has a bug.

**Comment by mubed** [comment]
> I see the null if you try a round trip. The null does not appear in the Format-Table output.
> 
> ```
> PS> Get-Content -LiteralPath file.json | ConvertFrom-Json | ConvertTo-Json
> [
>   null,
>   {
>     "key": "value"
>   }
> ]
> ```
> 
> and the Out-String solves the problem
> 
> ```
> PS> Get-Content -LiteralPath file.json | Out-String | ConvertFrom-Json | ConvertTo-Json
> {
>   "key": "value"
> }
> ```
> 
> So one could say that the issue is not one of ConvertFrom-Json parsing a single JSON document, it is part of the piping. This is the list of strings in the input pipeline to ConvertFrom-Json.
> 
> ```
> PS> Get-Content -LiteralPath file.json | ConvertTo-Json
> [
>   "// Comment on top level",
>   "{",
>   "\t\"key\": \"value\",",
>   "}"
> ]
> ```
> 
> The comment is on its own line so is a complete self contained JSON-with-comments document, it can be parsed in it's entirety and written to the output. The remaining lines are a group of lines referring to the same JSON document, so the parser has not seen the closing "}" so there is still more to come in subsequent lines.
> 
> So the ambiguity is two-fold,
> 
> 1. what counts as a complete JSON document
> 2. how are JSON documents represented as strings in the input pipeline
> 
> I suggest that ConvertFrom-JSON really sees two separate JSON documents.
> 
> If the comment is moved to within the brackets, the problem goes away
> 
> ```
> PS> Get-Content -LiteralPath file.json | ConvertFrom-Json | ConvertTo-Json
> {
>   "key": "value"
> }
> ```

Stil, ConvertFrom-Json is‌ designed to generate one single object representation (can also be an array) based on an array of string lines or a single multi line string. Parsing separate input items (JSON documents) which are not written as JSON arrays (wrapped in brackets, separated by commas) should not result in a JSON array, but a parsing error. ConverFrom-Json could simply ignore what comes after comment literals and add no data to the object structure during parsing.

**Comment by rhubarb-geek-nz** [comment]
Lets start with Get-Content and note that each line of a text file is its own record in the output pipeline.

I suggest that any change is going to make the matter worse. ConvertFrom-JSON does work with multiple JSON documents in different record records and outputs each as a record

With the file

```
{"key":"value"}
{"key":"value"}
{"key":"value"}
{"key":"value"}
{"key":"value"}
```

This is not a JSON file, it would need wrapping square brackets and comments between the objects. It is however a sequence of well-formed JSON files.

This is output as multiple PSCustomObjects

```
PS> Get-Content -LiteralPath file.json | ConvertFrom-Json | ForEach-Object { $_.GetType() }

IsPublic IsSerial Name                                     BaseType
-------- -------- ----                                     --------
True     False    PSCustomObject                           System.Object
True     False    PSCustomObject                           System.Object
True     False    PSCustomObject                           System.Object
True     False    PSCustomObject                           System.Object
True     False    PSCustomObject                           System.Object
```

And if you round trip them, then ConvertTo-JSON turns the list of records into a JSON array.

```
PS> Get-Content -LiteralPath file.json | ConvertFrom-Json | ConvertTo-Json
[
  {
    "key": "value"
  },
  {
    "key": "value"
  },
  {
    "key": "value"
  },
  {
    "key": "value"
  },
  {
    "key": "value"
  }
]
```

So ConvertFrom-JSON as it is today will read multiple JSON files and output each as a PSCustomObject

a single string containing a "// Comment on top level" is a complete self-contained JSON document as far as it is concerned, it had no parsing errors and gave back null, so that is what was written to the output pipeline.

If people are relying on the behaviour of ConvertFrom-JSON to parse each record as a unique JSON file then I don't see what change can be made to the cmdlet.

Now if you look at the document page it says

> You can pipe a JSON string to ConvertFrom-Json.

That suggests that is what you should be doing via Out-String or using Raw on Get-Content, where you are telling ConvertFrom-JSON what the boundaries of the document are, and there is exactly one in the input.

```
PS> Get-Content -LiteralPath file.json -Raw | ConvertFrom-Json | ConvertTo-Json
{
  "key": "value"
}
```

**Comment by rhubarb-geek-nz** [comment]
Showing each JSON document as its own record

```
PS> Get-Content -LiteralPath file.json
{"key": "value"}
42
"foo"
true
```

Then show the output of ConvertFrom-JSON

```
PS>  Get-Content -LiteralPath file.json | ConvertFrom-Json | ForEach-Object { $_.GetType() }

IsPublic IsSerial Name                                     BaseType
-------- -------- ----                                     --------
True     False    PSCustomObject                           System.Object
True     True     Int64                                    System.ValueType
True     True     String                                   System.Object
True     True     Boolean                                  System.ValueType
```

**Comment by jhoneill** [comment]
I'd class this as a bug and I can explain WHY , I think 

<img width="478" alt="image" src="https://github.com/PowerShell/PowerShell/assets/16209593/84435441-80c2-4789-a9b2-d43299b5e03f">

This rather odd bit of code with two break points in it feeds JSON into `ConvertFrom` 1 line at a time letting me single step through it in the debugger and  **critically**  I get output before all the text has been processed (i.e. Convert will send stuff down the pipeline to the next even if the input arrives very slowly, so object 1 gets processed before the last object has been sent. )  

I think when the text is sent line by line,  the initial comment means that when the next line is reached the process says "OK, starting a new object send the one we were working on" which is null.  That case was never tested for (and I'm not sure if that says a bug in PowerShell or in .NET stuff called by PowerShell).  
Processing the whole thing in one go processes correctly (and follows what the docs say about // being a comment)  

Incidentally   if you are reading from a file `convertfrom-Json (gc -raw foo.json)`  not only works around this problem but should be faster (although not noticeably unless files are very large).   If the JSON is arriving a little at a time from something else , piping is going to be better.

**Comment by mubed** [comment]
> Showing each JSON document as its own record
> 
> ```
> PS> Get-Content -LiteralPath file.json
> {"key": "value"}
> 42
> "foo"
> true
> ```
> 
> Then show the output of ConvertFrom-JSON
> 
> ```
> PS>  Get-Content -LiteralPath file.json | ConvertFrom-Json | ForEach-Object { $_.GetType() }
> 
> IsPublic IsSerial Name                                     BaseType
> -------- -------- ----                                     --------
> True     False    PSCustomObject                           System.Object
> True     True     Int64                                    System.ValueType
> True     True     String                                   System.Object
> True     True     Boolean                                  System.ValueType
> ```
> 
> Also a comment by itself is a complete JSON document, giving you a null, the following gives no parse errors.
> 
> ```
> PS> "// comment" | ConvertFrom-JSON
> ```
> 
> the output record is there....
> 
> ```
> PS> "// comment" | ConvertFrom-JSON | ForEach-Object { "." }
> .
> ```
> 
> just like
> 
> ```
> PS> "true" | ConvertFrom-JSON
> True
> ```
> 
> Now if you try
> 
> ```
> // comment
> {"key": "value"}
> 42
> "foo"
> true
> ```
> 
> You get in a mess
> 
> ```
> PS>  Get-Content -LiteralPath file.json | ConvertFrom-Json
> ConvertFrom-Json: Conversion from JSON failed with error: Additional text encountered after finished reading JSON content: 4. Path '', line 3, position 0.
> ```
> 
> So although "// comment" was sufficient to be a document by itself, when that was a prefix it assumed the following was all one document. So there looks like there is some behaviour to try and join the individual lines into a single document.

According to RFC7159 you are right. According to the older and more common understanding of what a JSON file should be (only object or array), you are wrong.

I would recommend to add a note to the documentation to raise awareness for this gotcha. Something like "always read the file in raw mode".

Another suggestion is, to introduce a new CmdLet called Import-Json analogous to Import-Csv that implicitly does the same thing.

**Comment by rhubarb-geek-nz** [comment]
> According to RFC7159 you are right. According to the older and more common understanding of what a JSON file should be (only object or array as in RFC4627), you are wrong.

I was just trying to describe the existing behaviour and it happily parses individual elements, so let's say ConvertFrom-JSON is consistent with RFC7159 from the point of view of what constitutes a complete JSON document.

```
PS> 'true' | ConvertFrom-JSON
True
```

But of course, neither of RFC7159 or RFC4627 describe any comment encoding.

**Comment by mklement0** [comment]
Let me try to provide a summary:

* **Context**:

  * Indeed, PowerShell Core's JSON parsing is based on Newtonsoft.Json; the latter supports not only _tolerating_ comments, but also _preserving_ them (_inside_ a JSON text), but the intent of `ConvertFrom-Json` is to _ignore_ them.

    * #14553 is a separate bug: _inside arrays_, comments are unexpectedly not only not ignored, but mistakenly turned into regular string array elements (I've posted a detailed analysis in https://github.com/PowerShell/PowerShell/issues/14553#issuecomment-1937770227)

  * While moving the implementation to `System.Text.Json` was considered, the relevant PR (#11198) was abandoned and is closed; more importantly, it would lead to breaking changes that would be considered unacceptable (loss of support for single-quoted property names and values, loss of support for unquoted property names).
  
    * Either way, taking away support for quietly tolerating comments would  _not_ be acceptable (`System.Text.Json` _does_ have support for that, but not for single-quoting / unquoted property names).

  * `ConvertFrom-Json` has a _heuristic_ built in, based on whether the first line of input is a self-contained (complete) JSON text (JSON document) by itself:

    * If it is, the line is parsed into an object (graph) by itself and the latter is output (streamed) right away; all subsequent lines must then be self-contained too and are processed the same.

    * If it is not, all input strings (lines) are _collected up front_, and only after all have been received are the lines _together_ parsed as a single JSON text.

* **The bug at hand**:

  * The bug at hand as well as #18804 are variations of the one discussed in #12229.

* **Already implemented / suggested _mitigations_**:

    * [Implemented] Example 3 in the [docs](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/convertfrom-json?view=powershell-7.4#example-3-convert-a-json-string-to-a-custom-object) now shows use of `Get-Content -Raw` - albeit with little conceptual explanation; the heuristic deserves explication in the "Description" section.

   * [Suggested, up for grabs] https://github.com/PowerShell/PSScriptAnalyzer/issues/1951

* **Suggestion for a - hopefully - [bucket 3](https://github.com/PowerShell/PowerShell/blob/master/docs/dev-process/breaking-change-contract.md#bucket-3-unlikely-grey-area) _fix_** (i.e. a change that is technically breaking, but acceptable for its benefits and presumed unlikeliness to affect existing) code (evolved from https://github.com/PowerShell/PowerShell/issues/12229#issuecomment-874319626):

  * When input strings (lines) are parsed _one at a time_, do _not_ emit a `$null` for, i.e. _effectively ignore_:
    * an _empty_ or _blank_ (all-whitespace) input string 
    * a _comment-only_ string, i.e. a string (line) composed solely of either `// ...` or  `/* ... */` (aside from any leading or trailing whitespace)
    * Rationale: 
      * Comments are fundamentally not part of the JSON standard, and an empty or blank string _in isolation_ is _not_ a valid JSON text (as defined in [RFC 8259](https://www.rfc-editor.org/rfc/rfc8259)).
      * Given that such strings are ignored _inside_ JSON texts, it stands to reason to also ignore them _between_ them.


  * When such strings (lines) are encountered at the _start_ of the input (causing them to be ignored with respect to _output_), _defer_ applying the heuristic until the first input string (line) that is neither blank/empty nor a comment-only string is encountered:

      * If that string is a self-contained JSON text, `ConvertFrom-Json` should also ignore any _subsequent_ empty/blank/comment-only lines, as implied by the first bullet point; otherwise, processing is as before, _except_:
        * If a malformed/incomplete string (line) is encountered, a _non_-terminating error must be reported, so that the command runs to completion and still outputs results for the _well-formed_ strings among the input. (The currently used (statement-)_terminating_ error in effect prevents _any_ output in _programmatic_ processing, though _on the screen only_ you will see results for the strings processed _before_ such an error occurs.)

      * If not, i.e. if collect-all-(remaining)-input-strings-first behavior is triggered, `ConvertFrom-Json` can rely on the underlying Newtonsoft.Json API to remove incidental whitespace and comments (including _in-line_ comments), which it _mostly_ already does, except for the JSON _array_-related bug reported in https://github.com/PowerShell/PowerShell/issues/14553#issuecomment-1937770227

  * Implementing the above should resolve all of the issues mentioned, including the one at hand.
